### ISSUE TITLE
Countries not added when adding a new product through the app

### ISSUE THREAD

**Issue by Ban3** [openfoodfacts/openfoodfacts-server#4147]
**Describe the bug**
Filing this against the server, because the app on my phone has been the same version for months (3.2.8, latest from F-Droid), but this only started happening recently.

When I add a new product with the Android app, the countries field is not set automatically like it used to be. Some example products I've added: [8410525220241](https://world.openfoodfacts.org/product/8410525220241/triss-i-hjarter-fini), [6408790001148](https://world.openfoodfacts.org/product/6408790001148/paistettu-lohipyorykka-hatala)

Notice how the addition line in the history says:

```txt
September 3, 2020 at 9:35:51 PM CEST - jumati Data (Added: lang) - Creating product (image upload)
```

Comparing it to older products:

```txt
August 25, 2020 at 8:44:36 PM CEST - jumati Data (Added: lang, countries) - Creating product (image upload)
```

I tried reproducing this with docker against current master with minimal product addition:

```sh
curl --request POST -d 'code=6438207019688&countries=en:finland' 'http://fi.productopener.localhost/cgi/product_jqm2.pl'
```

but that just added the product correctly with the countries set.

**Number of products impacted**
Is there way to search products where countries are unset? I looked at [products with Finnish barcodes without Finland in countries](https://world.openfoodfacts.org/code/64xxxxxxxxxxx/country/-finland)
and could see some where this is most likely happening (e.g [6438207019688](https://world.openfoodfacts.org/product/6438207019688/keiju-kaura) [6412000033775](https://world.openfoodfacts.org/product/6412000033775/halloumi-couscous-saarioinen)). Looking at [recently added products](https://world.openfoodfacts.org/entry-date/2020-09/editor/-kiliweb) similarly scores some which seem to suffer from this (e.g [0205503089082](https://world.openfoodfacts.org/product/0205503089082/langue-entiere-bovine-auchan))

These come from different app versions, so I doubt it's connected.

**Comment by stephanegigandet** [comment]
Products.pm init_product() tries to use the IP address to assign a country when creating a product (unless the app sends a country or uses a country subdomain like fi.openfoodfacts.org ). Have you changed your IP? (perhaps a VPN or something).

**Comment by Ban3** [comment]
> Have you changed your IP? (perhaps a VPN or something).
This is the same mobile network I've used the whole time. It doesn't just affect me, though.

**Comment by Ban3** [comment]
Can't figure this out. Checked how the app creates the product and it's just `https://ssl-api.openfoodfacts.org/cgi/product_image_upload.pl` with `code`, `imagefield` and `imgupload_front` parameters. So it must be here: https://github.com/openfoodfacts/openfoodfacts-server/blob/d6951e938d75c929b5d69a3b640812701d8517bc/lib/ProductOpener/Products.pm#L620-L624
It having [recent changes](https://github.com/openfoodfacts/openfoodfacts-server/commit/60323469890bcffbec29b81487c922064de1f99b#diff-10c7f6bdb20fc76e059eeb734a7f85a2) makes it even more of a suspect.

But when stepping through this with debugger on my local docker installation everything seems to work fine. Trying the subroutine with my phone's IP works just fine:
```perl
  DB<1> require ProductOpener::GeoIP

  DB<2> x ProductOpener::GeoIP::get_country_for_ip(<IP>)
0  'Finland'
```

**Comment by Ban3** [comment]
Meanwhile I tried to do some ballpark estimates of number of products affected. I had the JSONL export on hand, so I wrote a [jq script](https://gist.github.com/Ban3/8aa0d2064cf5c5a7a12315082cfcbed0) to filter it to a somewhat arbitrary criteria:
- Product added between 2020-09-01 and 2020-09-10. The issue didn't appear exactly at the start of the month, but it gives a nice starting point.
- No countries set on the product
- Product has photos. Photo upload is how the official apps creates the product
- Not added by some specific 3rd party apps that add products differently.

Products added during the same period in the previous month are used for comparison.

From this we get 677 products in this month and 0 in the comparison group. Total number of products added during these periods are 6504 and 7271

**Comment by Ban3** [comment]
Okay, I think I figured it out. TL;DR: Reversed if statement. A bit more detailed explanation:

Relevant part of my docker network:

- `172.24.0.6`: backend 
- `172.24.0.7`: frontend
- `172.24.0.1`: host

While making requests to `product_image_upload.pl` and looking at the execution, I noticed that `$ENV{REMOTE_ADDR}` was set to `172.24.0.7`.
That seemed weird as I'd expect the host address there, which was correctly set in `$ENV{HTTP_X_FORWARDED_FOR}`.
Turning the focus to the parts that should set that IP, I noticed this never got run:

https://github.com/openfoodfacts/openfoodfacts-server/blob/fff1357f774ac92bdaf666233f5011b52ab69b69/lib/startup_apache2.pl#L93-L96

There was a [recent change](https://github.com/openfoodfacts/openfoodfacts-server/commit/d07505aee7516bdfef86e87fc780e88623de341d#diff-6966674883878699016085eb850ba307) in that subroutine changing a statement from `unless` to `if`, but the condition was never changed to match.

Negating the condition correctly sets `$ENV{REMOTE_ADDR}` to match `$ENV{HTTP_X_FORWARDED_FOR}`.

I completely missed this on my first try because I was overriding `$ENV{REMOTE_ADDR}` to have a public IP there :roll_eyes:
